[AIRFLOW-5231] Fix S3Hook.delete_objects method

shcherbin opened the pull request:
---
Issue link: [AIRFLOW-5231](https://issues.apache.org/jira/browse/AIRFLOW-5231)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

shcherbin committed:
[AIRFLOW-5231] delete objects in batches with S3 hook.

feluelle commented on airflow/providers/amazon/aws/hooks/s3.py:
```diff
@@ -606,11 +606,22 @@ def delete_objects(self, bucket, keys):
+                )
+            except ClientError as e:
+                self.log.error(e.response["Error"]["Message"])
+                return False
```
I am not sure if we should return False on any `ClientError`. Can you be more specific?

feluelle replied:
And maybe then add a test to validate your `try-except` feature :) It should be a noted.

shcherbin replied:
I've changed the logic a bit due to failing tests in S3DeleteObjectsOperator.
ClientError is raised in case of any technical error (e.g. the specified bucked does not exist or the request is throttled).
I've also updated the tests.

shcherbin replied:
Ok, I don't see any other method in the code that deals with such technical issues. As long as the task will fail on any exception, it's fine with me.
Removing try-except.

shcherbin committed:
[AIRFLOW-5231] Update S3DeleteObjectsOperator due to api changes in S3 hook.

[AIRFLOW-5231] Update S3DeleteObjectsOperator due to api changes in S3 hook.

shcherbin committed:
[AIRFLOW-5231] Fix linting issue.

[AIRFLOW-5231] covert UT to pytest to avoid linting errors.

[AIRFLOW-5231] rename fixture to avoid name over-shadowing

mik-laj commented on airflow/providers/amazon/aws/hooks/s3.py:
```diff
@@ -610,7 +610,15 @@ def delete_objects(self, bucket, keys):
-        return response
+        s3 = self.get_conn()
+        batch = 1000
+        for i in range(0, len(keys), batch):
```
Why do you think about using airflow.helpers.chunks here?

shcherbin replied:
I didn't know that one existed. Thanks! :)

mik-laj commented on airflow/providers/amazon/aws/hooks/s3.py:
```diff
@@ -610,7 +610,15 @@ def delete_objects(self, bucket, keys):
-
-        return response
+        s3 = self.get_conn()
+        batch = 1000
```
Can you create a constant and add a comment to explain the reason for this change?

shcherbin replied:
I've added a comment and used the chunks method liked you asked. But I'm not going to make a constant out of the batch size, as it's an implementation detail specific only to the `S3Hook.delete_objects` method and will not be re-used elsewhere.

mik-laj commented on tests/providers/amazon/aws/hooks/test_s3.py:
```diff
@@ -0,0 +1,65 @@
+# under the License.
+#
+
+import pytest
```
Can you move [this file](https://github.com/apache/airflow/blob/master/tests/hooks/test_s3_hook.py) also?

shcherbin replied:
Sure.

shcherbin replied:
Done

shcherbin committed:
[AIRFLOW-5231] review: use utils.helpers.chunks; add a comment describing 'magic number'.

shcherbin committed:
[AIRFLOW-5231] review: migrate tests/hooks/test_s3_hook to tests/providers/amazon/aws/hooks/test_s3; covert asserts to pytest; use fixtures.

[AIRFLOW-5231] review: migrate tests/hooks/test_s3_hook to tests/providers/amazon/aws/hooks/test_s3; covert asserts to pytest; use fixtures.

feluelle commented on airflow/providers/amazon/aws/hooks/s3.py:
```diff
@@ -610,7 +611,18 @@ def delete_objects(self, bucket, keys):
-        response = self.get_conn().delete_objects(Bucket=bucket, Delete=delete_dict)
-
-        return response
+        s3 = self.get_conn()
+
+        # We can only send a maximum of 1000 keys per request.
+        # For details see:
+        # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.delete_objects
+        for chunk in chunks(keys, chunk_size=1000):
+            response = s3.delete_objects(
+                Bucket=bucket,
+                Delete={"Objects": [{"Key": k} for k in chunk]}
+            )
+            deleted_keys = [x['Key'] for x in response.get("Deleted", [])]
+            self.log.info("Deleted: %s", deleted_keys)
+            if "Errors" in response:
+                errors_keys = [x['Key'] for x in response.get("Errors", [])]
+                raise AirflowException("Errors when deleting: {}".format(errors_keys))
```
LGTM, but please mention in the Updating.md that you removed the return statement so that people are aware of this when updating :)

Some users might have implemented their own operator based on this `delete_objects` method.

shcherbin replied:
Done

shcherbin committed:
[AIRFLOW-5231] review: add a note to UPDATING.md stating the interface change.
